rust/src/operations/delete.rs

```diff
@@ -0,0 +1,944 @@
+                limit: None,
+                table_partition_cols,
+                infinite_source: false,
+                output_ordering: None,
+            },
+            None,
+        )
+        .await?;
+
+    let filter: Arc<dyn ExecutionPlan> =
+        Arc::new(FilterExec::try_new(predicate_expr, parquet_scan.clone())?);
+    let limit: Arc<dyn ExecutionPlan> = Arc::new(LocalLimitExec::new(filter, 1));
+
+    let task_ctx = Arc::new(TaskContext::from(state));
+    let partitions = limit.output_partitioning().partition_count();
+    let mut tasks = Vec::with_capacity(partitions);
+
+    for i in 0..partitions {
+        let stream = limit.execute(i, task_ctx.clone())?;
+        tasks.push(handle_stream(stream));
+    }
+
+    for res in futures::future::join_all(tasks).await.into_iter() {
+        let path = res?;
+        if let Some(path) = path {
+            match candidate_map.remove(&path) {
+                Some(action) => files.push(action),
+                None => {
+                    return Err(DeltaTableError::Generic(
+                        "Unable to map __delta_rs_path to action.".to_owned(),
+                    ))
+                }
+            }
+        }
+    }
+
+    Ok(files)
+}
+
+async fn handle_stream(
+    mut stream: Pin<Box<dyn RecordBatchStream + Send>>,
+) -> Result<Option<String>, DeltaTableError> {
+    if let Some(maybe_batch) = stream.next().await {
+        let batch: RecordBatch = maybe_batch?;
+        if batch.num_rows() > 1 {
+            return Err(DeltaTableError::Generic(
+                "Find files returned multiple records for batch".to_owned(),
+            ));
+        }
+        let array = batch
+            .column_by_name(PATH_COLUMN)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .ok_or(DeltaTableError::Generic(format!(
+                "Unable to downcast column {}",
+                PATH_COLUMN
+            )))?;
+
+        let path = array
+            .into_iter()
+            .next()
+            .unwrap()
+            .ok_or(DeltaTableError::Generic(format!(
+                "{} cannot be null",
+                PATH_COLUMN
+            )))?;
+        return Ok(Some(path.to_string()));
+    }
+
+    Ok(None)
+}
+
+struct ExprProperties {
+    partition_only: bool,
+}
+
+/// Ensure only expressions that make sense are accepted, check for
+/// non-deterministic functions, and determine if the expression only contains
+/// partition columns
+fn validate_expr(
+    expr: &Expr,
+    partition_columns: &Vec<String>,
+    properties: &mut ExprProperties,
+) -> DeltaResult<()> {
+    // TODO: We can likely relax the volatility to STABLE. Would require further
+    // research to confirm the same value is generated during the scan and
+    // rewrite phases.
+    match expr {
+        Expr::ScalarVariable(_, _) | Expr::Literal(_) => (),
+        Expr::Alias(expr, _) => validate_expr(expr, partition_columns, properties)?,
```

Comment:
Perhaps we can avoid recursion using the `TreeNode` trait methods on `Expr`?
https://docs.rs/datafusion/23.0.0/datafusion/prelude/enum.Expr.html#impl-TreeNode-for-Expr

Would help avoid stack overflows if there is a very nested expression. I could imagine that happening if someone passes a filter like:

```sql
x = 1
OR x = 2
OR x = 3
...
```

Reply:
The Datafusion implementation still uses recursion but overflow issues are now on them. The visitor pattern also makes the code look cleaner too!

Reply:
> The Datafusion implementation still uses recursion but overflow issues are now on them

🤣